code/controllers/subsystems/machines.dm

```diff
@@ -110,67 +118,111 @@ datum/controller/subsystem/machines/proc/setup_atmos_machinery(list/machines)
+			thing.is_processing = null
+			nextProcessingListPosition-- //removing processed thing from list moves the queue to the left, adjust accordingly
+
+		if(!(nextProcessingListPosition%tickCheckPeriod)) //pauses every tickCheckPeriod-th processed thing
+			pause()
 			return
 
+	nextProcessingListPosition = 0 //entire list was processed
+
 /datum/controller/subsystem/machines/proc/process_machinery(resumed = 0)
 	if (!resumed)
-		src.current_run = machinery.Copy()
-
-	var/list/current_run = src.current_run
-	while(current_run.len)
-		var/obj/machinery/M = current_run[current_run.len]
-		current_run.len--
-		if(istype(M) && !QDELETED(M) && !(M.Process(wait) == PROCESS_KILL))
-			if(M.use_power)
-				M.auto_use_power()
+		nextProcessingListPosition = 1 //fresh start, otherwise from saved posisition
+
+	//localizations
+	var/list/local_list = machinery
+	var/obj/machinery/thing
+	var/wait = src.wait
+
+	var/tickCheckPeriod = round(local_list.len/16+1) //pause process every 1/16th length of list
+	while(nextProcessingListPosition && (nextProcessingListPosition <= local_list.len)) //until position is valid
+		thing = local_list[nextProcessingListPosition]
+		nextProcessingListPosition++
+
+		if(!QDELETED(thing) && (thing.Process(wait) != PROCESS_KILL))
+			if(thing.use_power)
+				thing.auto_use_power()
 		else
-			machinery.Remove(M)
-			M.is_processing = null
-		if(MC_TICK_CHECK)
+			local_list.Remove(thing)
+			thing.is_processing = null
+			nextProcessingListPosition-- //removing processed thing from list moves the queue to the left, adjust accordingly
+
+		if(!(nextProcessingListPosition%tickCheckPeriod)) //pauses every tickCheckPeriod-th processed thing
+			pause()
 			return
 
+	nextProcessingListPosition = 0 //entire list was processed
+
 /datum/controller/subsystem/machines/proc/process_powernets(resumed = 0)
 	if (!resumed)
-		src.current_run = powernets.Copy()
-
-	var/list/current_run = src.current_run
-	while(current_run.len)
-		var/datum/powernet/PN = current_run[current_run.len]
-		current_run.len--
-		if(istype(PN) && !QDELETED(PN))
-			PN.reset(wait)
+		nextProcessingListPosition = 1 //fresh start, otherwise from saved posisition
+
+	//localizations
+	var/list/local_list = powernets
+	var/datum/powernet/thing
+	var/wait = src.wait
+
+	var/tickCheckPeriod = round(local_list.len/16+1) //pause process every 1/16th length of list
+	while(nextProcessingListPosition && (nextProcessingListPosition <= local_list.len)) //until position is valid
+		thing = local_list[nextProcessingListPosition]
+		nextProcessingListPosition++
+
+		if(!QDELETED(thing))
+			thing.reset(wait)
 		else
-			powernets.Remove(PN)
-			PN.is_processing = null
-		if(MC_TICK_CHECK)
+			local_list.Remove(thing)
+			thing.is_processing = null
+			nextProcessingListPosition-- //removing processed thing from list moves the queue to the left, adjust accordingly
+
+		if(!(nextProcessingListPosition%tickCheckPeriod)) //pauses every tickCheckPeriod-th processed thing
+			pause()
```

Comment:
Are you sure this pausing is beneficial? What's the goal of it?

Reply:
Sacrifices performance due to rescheduling for multiple smaller process calls, which makes for smaller perceived lag spikes.